### PR TITLE
732 add augment bounding boxes support to random translation layer

### LINKED ISSUE #732: Add augment_bounding_boxes support to RandomTranslation layer
The augment_bounding_boxes should be implemented for RandomTranslation Layer in keras_cv. The PR should contain implementation, test scripts and a demo script to verify implementation.

Example code for implementing augment_bounding_boxes() can be found here

- https://github.com/keras-team/keras-cv/blob/master/keras_cv/layers/preprocessing/random_flip.py#:~:text=def%20augment_bounding_boxes(,)%3A
- https://github.com/keras-team/keras-cv/blob/master/keras_cv/layers/preprocessing/random_rotation.py#:~:text=def%20augment_image(self%2C%20image%2C%20transformation%2C%20**kwargs)%3A
- The implementations can be verified using demo utils in keras_cv.bounding_box - Example of demo script can be found here : https://github.com/keras-team/keras-cv/blob/master/examples/layers/preprocessing/bounding_box/random_rotation_demo.py

### PR DESCRIPTION
**Pull request by akucia**
# What does this PR do?
Add augment_bounding_boxes to RandomTranslation layer.
Fixes #732 

## Example results (fill_mode=constant)
Figure_1
## Example results (fill_mode=reflect)
RandomTranslation_reflect

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [x] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

## Who can review?
@divyashreepathihalli @LukeWood

### CODE REVIEW

**Comment by LukeWood** [comment]
@akucia can you paste the results from the example script into the PR description?

File: keras_cv/layers/preprocessing/random_translation.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+            fill_mode=self.fill_mode,
+            fill_value=self.fill_value,
+            interpolation=self.interpolation,
+        )
+        output = tf.squeeze(output, 0)
+        output.set_shape(image_shape)
+        return output
+
+    @staticmethod
+    def _translate_bounding_boxes_horizontal(bounding_boxes, dx):
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/random_translation.py:162]: I think this code may be easier to follow if we perform horizontal and vertical translation together in one method, e.g:

```
 def _translate_bounding_boxes_vertical(bounding_boxes, dy):
        x0, y0, x1, y1, rest = tf.split(
            bounding_boxes, [1, 1, 1, 1, bounding_boxes.shape[-1] - 4], axis=-1
        )
        output = tf.stack(
            [
                x0 + dx,
                y0 + dy,
                x1 + dx,
                y1 + dy,
                rest,
            ],
            axis=-1,
        )
        output = tf.squeeze(output, axis=1)
        return output
```

File: keras_cv/layers/preprocessing/random_translation.py [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+    def augment_bounding_boxes(
+        self, bounding_boxes, transformation=None, image=None, **kwargs
+    ):
+        if self.bounding_box_format is None:
+            raise ValueError(
+                "`RandomTranslation()` was called with bounding boxes,"
+                "but no `bounding_box_format` was specified in the constructor."
+                "Please specify a bounding box format in the constructor. i.e."
+                "`RandomFlip(bounding_box_format='xyxy')`"
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/random_translation.py:205]: RandomTranslation

File: keras_cv/layers/preprocessing/random_translation.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+            target=self.bounding_box_format,
+            dtype=self.compute_dtype,
+            images=image,
+        )
+        return bounding_boxes
+
+    def augment_segmentation_mask(
+        self, segmentation_mask, transformation=None, **kwargs
+    ):
+        return self._translate_image(segmentation_mask, transformation)
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/random_translation.py:240]: When transforming a segmentation mask, we'll need to use `interpolation="nearest"` because segmentation masks cannot be interpolated numerically like pixels can (this could lead to erroneous values in the mask, e.g. if interpolating between class 0 and 6, you might compute class 3 which would be incorrect)

Comment (ianstenbit) [reply keras_cv/layers/preprocessing/random_translation.py:240]: (See a similar approach in [RandomResizedCrop](https://github.com/keras-team/keras-cv/blob/master/keras_cv/layers/preprocessing/random_resized_crop.py))

File: keras_cv/layers/preprocessing/random_translation_test.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+# limitations under the License.
+
+import numpy as np
+import tensorflow as tf
+from absl.testing import parameterized
+
+from keras_cv.layers.preprocessing.random_translation import RandomTranslation
+
+
+class RandomTranslationTest(tf.test.TestCase, parameterized.TestCase):
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/random_translation_test.py:22]: It looks like these tests currently only cover `fill_mode="constant"`.
Let's add some cases for reflect, nearest, and wrap.

Here's some example cases I wrote up (which you'll need to adapt a bit because I was using a different naming convention for the factor params):

```
    def test_translation_with_reflect(self):
        input_image = np.reshape(np.arange(0, 25), (1, 5, 5, 1))
        # Shift by 2 pixels
        layer = preprocessing.RandomTranslation(
            height_factor=(-0.4, -0.4), width_factor=(0, 0), fill_mode="reflect"
        )
        output_image = layer(input_image)
        expected_output = np.reshape(
            np.asarray(
                [
                    [10, 11, 12, 13, 14],
                    [15, 16, 17, 18, 19],
                    [20, 21, 22, 23, 24],
                    [20, 21, 22, 23, 24],
                    [15, 16, 17, 18, 19],
                ]
            ),
            (1, 5, 5, 1),
        )
        self.assertAllEqual(expected_output, output_image)

    def test_translation_with_constant(self):
        input_image = np.reshape(np.arange(0, 25), (1, 5, 5, 1))
        # Shift by 2 pixel
        layer = preprocessing.RandomTranslation(
            height_factor=(-0.4, -0.4),
            width_factor=(0, 0),
            fill_mode="constant",
            fill_value=0,
        )
        output_image = layer(input_image)
        expected_output = np.reshape(
            np.asarray(
                [
                    [10, 11, 12, 13, 14],
                    [15, 16, 17, 18, 19],
                    [20, 21, 22, 23, 24],
                    [0, 0, 0, 0, 0],
                    [0, 0, 0, 0, 0],
                ]
            ),
            (1, 5, 5, 1),
        )
        self.assertAllEqual(expected_output, output_image)

    def test_translation_with_wrap(self):
        input_image = np.reshape(np.arange(0, 25), (1, 5, 5, 1))
        # Shift by 1 pixel
        layer = preprocessing.RandomTranslation(
            height_factor=(-0.4, -0.4), width_factor=(0, 0), fill_mode="wrap"
        )
        output_image = layer(input_image)
        expected_output = np.reshape(
            np.asarray(
                [
                    [10, 11, 12, 13, 14],
                    [15, 16, 17, 18, 19],
                    [20, 21, 22, 23, 24],
                    [0, 1, 2, 3, 4],
                    [5, 6, 7, 8, 9],
                ]
            ),
            (1, 5, 5, 1),
        )
        self.assertAllEqual(expected_output, output_image)

    def test_translation_with_nearest(self):
        input_image = np.reshape(np.arange(0, 25), (1, 5, 5, 1))
        # Shift by 2 pixels
        layer = preprocessing.RandomTranslation(
            height_factor=(-0.4, -0.4), width_factor=(0, 0), fill_mode="nearest"
        )
        output_image = layer(input_image)
        expected_output = np.reshape(
            np.asarray(
                [
                    [10, 11, 12, 13, 14],
                    [15, 16, 17, 18, 19],
                    [20, 21, 22, 23, 24],
                    [20, 21, 22, 23, 24],
                    [20, 21, 22, 23, 24],
                ]
            ),
            (1, 5, 5, 1),
        )
        self.assertAllEqual(expected_output, output_image)
```

File: keras_cv/utils/preprocessing.py [hunk 5]
```diff
@@ -94,13 +94,45 @@ def blend(image1: tf.Tensor, image2: tf.Tensor, factor: float) -> tf.Tensor:
...
+    if range_mode not in allowed_range_modes:
+        raise ValueError(
+            f"`range_mode` should be one of {allowed_range_modes}. "
+            f"Got range_mode={range_mode}"
+        )
     if isinstance(param, core.FactorSampler):
         return param
 
-    if isinstance(param, float) or isinstance(param, int):
+    if isinstance(param, (float, int)) and range_mode == "non-negative":
```
Comment (ianstenbit) [inline keras_cv/utils/preprocessing.py:130]: Calling this mode "non-negative" seems to assume that min_value is 0, which is not enforced by this API. I can't think of a better name for this, though, and I like the idea of including this as an option when parsing factors.

@LukeWood WDYT

File: keras_cv/layers/preprocessing/random_translation.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+                x3,
+                x4 + dy,
+                rest,
+            ],
+            axis=-1,
+        )
+        output = tf.squeeze(output, axis=1)
+        return output
+
+    def augment_bounding_boxes(
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/random_translation.py:197]: As implemented, this works as-intended for `fill_mode="constant"` (and I think "nearest"), but not for "reflect" and "wrap".

When wrapping or reflecting, we'll need to create new bounding boxes (or expand existing ones) for the reflected or wrapped pixels that were previously inside of a bounding box (and then potentially prune the new ones if they are unreasonably small).

Here's an example I produced from running your example script:
Screen Shot 2022-09-22 at 10 54 26 AM

Notice that the mirrored cow on the right does not have a bounding box on it (but it should).

If you'd like to, for now we can raise an exception if this layer is called on bounding boxes with one of the non-constant fill modes, and leave this code as-is. Otherwise, let's update it to support the other fill modes.

**Review by ianstenbit — COMMENTED** [review]
Thank you for the PR!

This is great -- we're excited about getting RandomTranslation support set up for bounding boxes, as this is a big step toward supporting RandAugment for bboxes.

File: keras_cv/utils/preprocessing.py (hunk 5 above)
Comment (ianstenbit) [reply keras_cv/utils/preprocessing.py:130]: Yeah that sounds reasonable to me

File: keras_cv/layers/preprocessing/random_translation.py (hunk 6 above)
Comment (akucia) [reply keras_cv/layers/preprocessing/random_translation.py:197]: I tried to check how other frameworks handle that (ie albumentations) and I don't think they create new bounding boxes. There's also a case, when reflected box' class might no longer be a valid. Anyway I think you're right and it should be added.

Comment (LukeWood) [reply keras_cv/layers/preprocessing/random_translation.py:179]: we should raise an error if you’re augmenting bounding boxes with reflect or wrap fill mode until we implement a projective transform for bounding boxes IMO

File: keras_cv/layers/preprocessing/random_translation.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+            source=self.bounding_box_format,
+            target="rel_xyxy",
+            images=image,
+        )
+        bounding_boxes = RandomTranslation._translate_bounding_boxes(
+            bounding_boxes,
+            transformation["translate_horizontal"],
+            transformation["translate_vertical"],
+        )
+
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/random_translation.py:201]: Can we add a unit test covering the case where a bounding box is translated fully out of the image?

File: keras_cv/layers/preprocessing/random_translation.py [hunk 8]
```diff
@@ -193,17 +299,24 @@ def augment_bounding_boxes(
...
             transformation["translate_vertical"],
+            self.fill_mode,
         )
 
         bounding_boxes = bounding_box.clip_to_image(
             bounding_boxes,
             bounding_box_format="rel_xyxy",
             images=image,
         )
+        bounding_boxes = bounding_box.reject_empty(
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/random_translation.py:314]: Thank you for this API; I actually REALLY like this API.

**Comment by akucia** [comment]
There's still some things left to do:

- [ ] test case with various number of bboxes per image

**Comment by akucia** [comment]
There's still stuff to do:
- [ ] I want to add a test case with various number of bboxes per image
- [ ] test pass, but demo is failing, maybe it's related to the point above
- [ ] `_transform_bounding_boxes` is messy, I think I can make it simpler
- [ ] resolve conflicts

File: keras_cv/layers/preprocessing/random_translation.py (hunk 8 above)
Comment (akucia) [reply keras_cv/layers/preprocessing/random_translation.py:314]: @LukeWood I implemented this in 2 ways: 1st with ragged tensor outputs and 2nd with masking with sentinel values. I don't know which one would be preferred, but the ragged tensors didn't work with other elements of the data pipeline.

**Comment by akucia** [comment]
@LukeWood @ianstenbit  I have a problem using this layer with `RandomAugmentationPipeline`. I get the following errors:
```
2022-10-03 22:01:34.819398: W tensorflow/core/framework/op_kernel.cc:1780] OP_REQUIRES failed at list_kernels.h:206 : INVALID_ARGUMENT: PartialTensorShape: Incompatible shapes during merge: [20,5] vs. [80,5]
2022-10-03 22:01:34.820795: W tensorflow/core/framework/op_kernel.cc:1780] OP_REQUIRES failed at list_kernels.h:206 : INVALID_ARGUMENT: PartialTensorShape: Incompatible shapes during merge: [16,5] vs. [4,5]
...
```
Do you have any ideas what's not working correctly? 

Here's a code snippet to reproduce this:
```
import keras_cv
import tensorflow as tf
from examples.layers.preprocessing.bounding_box import demo_utils

AUGMENTATION_LAYERS = [
    keras_cv.layers.preprocessing.RandomFlip(bounding_box_format="xywh"),
    keras_cv.layers.preprocessing.RandomTranslation(
        x_factor=0.5,
        y_factor=0.5,
        fill_mode="reflect",
        bounding_box_format="xywh",
    ),
]
pipeline = keras_cv.layers.RandomAugmentationPipeline(
    layers=AUGMENTATION_LAYERS, augmentations_per_image=2
)
dataset = demo_utils.load_voc_dataset(bounding_box_format="xywh")
augmented_dataset = dataset.map(pipeline, num_parallel_calls=tf.data.AUTOTUNE)
demo_utils.visualize_data(augmented_dataset, bounding_box_format="xywh")
```
What's strange is that using a single augmentations doesn't cause problems:
```
import keras_cv
import tensorflow as tf
from examples.layers.preprocessing.bounding_box import demo_utils

AUGMENTATION_LAYERS = [
    keras_cv.layers.preprocessing.RandomTranslation(
        x_factor=0.5,
        y_factor=0.5,
        fill_mode="reflect",
        bounding_box_format="xywh",
    ),
]
pipeline = keras_cv.layers.RandomAugmentationPipeline(
    layers=AUGMENTATION_LAYERS, augmentations_per_image=2
)
dataset = demo_utils.load_voc_dataset(bounding_box_format="xywh")
augmented_dataset = dataset.map(pipeline, num_parallel_calls=tf.data.AUTOTUNE)
demo_utils.visualize_data(augmented_dataset, bounding_box_format="xywh")
```

**Comment by divyashreepathihalli** [comment]
Hey @akucia, thank you for the PR. Do you have any updates on this?

**Comment by LukeWood** [comment]
Temporarily closing due to inactivity.  Re-open if you'd like to push this forward
